### PR TITLE
Fix escope to take sourceType and ecmaVersion from options

### PR DESCRIPTION
**Pull request by danez**
escope was hardcoded to sourcetype: "module" and ecmaVersion: "6"
This changes it to take the configuration from the eslint options and still
defaulting to "module" and "6".
This is done by having the variables in the upper scope, as monkeypatch is only triggered once and would therefore stick to the options that are supplied the first time. But the options can change as in our tests or different eslintrc files in different directories.

Fixes #287 and #217 (217 was failing without the fix in script mode)

### CODE REVIEW

**Comment by josh** [comment]
Nice! Thanks @danez

File: test/non-regression.js [hunk 1]
```diff
@@ -1362,6 +1367,54 @@ describe("verify", function () {
...
+      [ "1:5 Implicit global variable, assign as global property instead. no-implicit-globals" ],
+      "script",
+      {
+        env: {},
+        parserOptions: { ecmaVersion: 6, sourceType: "script" }
+      }
+    );
+  });
+
+  it("no-implicit-globals in module", function () {
```
Comment (hzoo) [inline test/non-regression.js]: I guess we should add a test for when the parserOptions aren't specified?

File: test/non-regression.js [hunk 2]
```diff
@@ -2,27 +2,32 @@
...
-      parserOptions: {
-        ecmaVersion: 6,
-        ecmaFeatures: {
-          jsx: true,
-          experimentalObjectRestSpread: true,
-          globalReturn: true
-        },
-        sourceType: sourceType || "module"
-      }
+      sourceType: sourceType || "module"
```
Comment (hzoo) [inline test/non-regression.js]: And for that test we may need to remove ` || "module"`?

**Comment by danez** [comment]
I removed the defaulting to `"module"` in the test as @hzoo suggested. Then some tests were failing as the default value for sourceType and ecmaVersion was not working correctly.

After fixing that now only one test fails, but I don't have time right now to investigate. Seems there is still a difference between setting `sourceType='module'` or not setting it.

**Comment by josh** [comment]
> After fixing that now only one test fails, but I don't have time right now to investigate. Seems there is still a difference between setting sourceType='module' or not setting it.

Looks like you're referring to the "no-use-before-define #192" test. It looks like @hzoo suggested modules must be used in this case.

**Comment by danez** [comment]
Ok I finally had time to look at this and figured out what the problem was with the "no-use-before-define" rule test.
The rule was checking the global scope and its first childScope for use-before-define breakages. But babel-eslint (at least in the tests) was setting nodejsScope = true and sourceType module, which created 2 childscopes, one for the nodjes-env and one for the module. And this somehow confused the rule.

Now the nodejsScope is et to false if sourceType is module. Like eslint is doing that: 
https://github.com/eslint/eslint/blob/5685ff58d74cb2b7096bcfa1dfc6cae95c2a9278/lib/eslint.js#L834
https://github.com/eslint/eslint/blob/5685ff58d74cb2b7096bcfa1dfc6cae95c2a9278/lib/eslint.js#L449

**Comment by josh** [comment]
This looks good to me!

**Comment by danez** [comment]
@hzoo Do you have time to look at this?

**Comment by josh** [comment]
Friendly bump 😄

**Comment by hzoo** [comment]
Sorry all - busy and maybe didn't want to check all of this 😄 .

**Comment by josh** [comment]
@hzoo ❤️

**Comment by hzoo** [comment]
Also @josh @danez Do yall want to be contributors? (ref to #88)

**Pull request merged**